File: Week6/sentimental-credit/credit.py
from cs50 import get_int
import math


def isValidA(creditNumber, num):
    digits = math.trunc(creditNumber / pow(10, 13))
    if ((num == 15) and (digits == 34 or digits == 37)):
        return True
    else:
        return False


def isValidV(creditNumber, num):
    if (num == 13):
        digits = math.trunc(creditNumber / pow(10, 12))
        if (digits == 4):
            return True
    elif num == 16:
        digits = math.trunc(creditNumber / pow(10, 15))
        if (digits == 4):
            return True
    return False


def isValidM(creditNumber, num):
    digits = math.trunc(creditNumber / pow(10, 14))
    if ((num == 16) and (digits > 50 and digits < 56)):
        return True
    else:
        return False


def digitcounter(creditNumber):
    count = 0
    while (creditNumber > 0):
        count += 1
        creditNumber = math.trunc(creditNumber / 10)
    return count


def productSum(lastDigit):
    multiply = lastDigit * 2
    sum = 0
    while (multiply > 0):
        answer = multiply % 10
        sum = sum + answer
        multiply = math.trunc(multiply / 10)
    return sum


def evenSum(creditNumber):
    sum = 0
    flag = False
    while (creditNumber > 0):
        if (flag == True):
            lastDigit = creditNumber % 10
            product = productSum(lastDigit)
            sum += product
        else:
            lastDigit = creditNumber % 10
            sum += lastDigit
        flag = not (flag)
        creditNumber = math.trunc(creditNumber / 10)
    return sum


def main():
    creditNumber = get_int("Enter your credit card number: ")
    eSum = evenSum(creditNumber)
    num = digitcounter(creditNumber)
    a = isValidA(creditNumber, num)
    m = isValidM(creditNumber, num)
    v = isValidV(creditNumber, num)
    if (not (eSum % 10 == 0)):
        print("INVALID")
        return 0
    elif (a == True):
        print("AMEX")
    elif (m == True):
        print("MASTERCARD")
    elif (v == True):
        print("VISA")
    else:
        print("INVALID")
        return 0


main()

File: Week1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>
int evenSum(long creditNumber);
int productSum(int lastDigit);
int digitCounter(long creditNumber);
bool isValidA(long creditNumber, int num);
bool isValidM(long creditNumber, int num);
bool isValidV(long creditNumber, int num);

int main(void)
{
    long creditNumber = get_long("Enter your credit card number: ");
    int eSum = evenSum(creditNumber);
    int num = digitCounter(creditNumber);
    bool a = isValidA(creditNumber, num);
    bool m = isValidM(creditNumber, num);
    bool v = isValidV(creditNumber, num);
    if (eSum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    else if (a == true)
    {
        printf("AMEX\n");
    }
    else if (m == true)
    {
        printf("MASTERCARD\n");
    }
    else if (v == true)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

bool isValidA(long creditNumber, int num)
{
    int digits = creditNumber / pow(10, 13);
    if ((num == 15) && (digits == 34 || digits == 37))
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool isValidV(long creditNumber, int num)
{
    if (num == 13)
    {
        int digits = creditNumber / pow(10, 12);
        if (digits == 4)
        {
            return true;
        }
    }
    else if (num == 16)
    {
        int digits = creditNumber / pow(10, 15);
        if (digits == 4)
        {
            return true;
        }
    }
    return false;
}

bool isValidM(long creditNumber, int num)
{
    int digits = creditNumber / pow(10, 14);
    if ((num == 16) && (digits > 50 && digits < 56))
    {
        return true;
    }
    else
    {
        return false;
    }
}

int digitCounter(long creditNumber)
{
    int count = 0;
    while (creditNumber > 0)
    {
        count++;
        creditNumber = creditNumber / 10;
    }
    return count;
}

int evenSum(long creditNumber)
{
    int sum = 0;
    bool flag = false;
    while (creditNumber > 0)
    {
        if (flag == true)
        {
            int lastDigit = creditNumber % 10;
            int product = productSum(lastDigit);
            sum += product;
        }
        else
        {
            int lastDigit = creditNumber % 10;
            sum += lastDigit;
        }
        flag = !flag;
        creditNumber = creditNumber / 10;
    }
    return sum;
}

int productSum(int lastDigit)
{
    int multiply = lastDigit * 2;
    int sum = 0;
    while (multiply > 0)
    {
        int answer = multiply % 10;
        sum = sum + answer;
        multiply = multiply / 10;
    }
    return sum;
}
